python_modules/dagster/dagster/_daemon/asset_daemon.py

```diff
@@ -51,6 +77,72 @@ def get_current_evaluation_id(instance: DagsterInstance) -> Optional[int]:
     return AssetDaemonCursor.get_evaluation_id_from_serialized(raw_cursor) if raw_cursor else None
 
 
+class AutoMaterializeLaunchContext:
+    def __init__(
+        self,
+        tick: InstigatorTick,
+        instance: DagsterInstance,
+        logger: logging.Logger,
+        tick_retention_settings,
+    ):
+        self._tick = tick
+        self._logger = logger
+        self._instance = instance
+
+        self._purge_settings = defaultdict(set)
+        for status, day_offset in tick_retention_settings.items():
+            self._purge_settings[day_offset].add(status)
+
+    @property
+    def status(self) -> TickStatus:
+        return self._tick.status
+
+    @property
+    def logger(self) -> logging.Logger:
+        return self._logger
+
+    def add_run_info(self, run_id=None):
+        self._tick = self._tick.with_run_info(run_id)
+
+    def update_state(self, status: TickStatus, **kwargs: object):
+        self._tick = self._tick.with_status(status=status, **kwargs)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(
+        self,
+        exception_type: Type[BaseException],
+        exception_value: Exception,
+        traceback: TracebackType,
+    ) -> None:
+        if exception_value and isinstance(exception_value, KeyboardInterrupt):
+            return
+
+        # Log the error if the failure wasn't an interrupt or the daemon generator stopping
+        if exception_value and not isinstance(exception_value, GeneratorExit):
```

Comment:
do we have the hooks to log this to DD in cloud?  This generally should not happen, right?

Reply:
yeah, failed ticks show up in datadog - an example of when it could happen in an 'expected' way currently is if one of your assets getting materialized references an environment variable that doesn't exist